src/framework/mpas_field_routines.F

```diff
@@ -1716,44 +1869,66 @@ end subroutine mpas_deallocate_field1d_real!}}}
 !> \date    04/02/13
 !> \details 
-!> This routine deallocates a 2D real field.
+!>  This routine deallocates a 2D real field. If the optional argument
+!>  keep_head_ptr is .true., the field type pointed to by f will not be
+!>  deallocated -- only the contents (array, attLists, constituentNames)
+!>  will be deallocated. If the field contains multiple blocks, all memory
+!>  for all blocks other than the first (including the field types themselves)
+!>  will be deallocated.
+!>
+!>  The principal use fo rthe keep_head_ptr argument is for situations in
+!>  which elemements of an array of fields need to be deallocated.
 !
 !-----------------------------------------------------------------------
-   subroutine mpas_deallocate_field2d_real(f)!{{{
-       type (field2dReal), pointer :: f !< Input: Field to deallocate
-       type (field2dReal), pointer :: f_cursor
+   subroutine mpas_deallocate_field2d_real(f, keep_head_ptr)!{{{
+
+       implicit none
+
+       type (field2dReal), target :: f !< Input: Field to deallocate
+       logical, intent(in), optional :: keep_head_ptr
+
+       type (field2dReal), pointer :: f_cursor, f_next
+       logical :: local_keep_head_ptr
        integer :: threadNum
        integer :: i, iErr
 
        threadNum = mpas_threading_get_thread_num()
 
+       if (present(keep_head_ptr)) then
+           local_keep_head_ptr = keep_head_ptr
+       else
+           local_keep_head_ptr = .false.
+       end if
+
        if ( threadNum == 0 ) then
-          f_cursor => f
-          do while(associated(f_cursor))
-            if(associated(f % next)) then
-              f => f % next
-            else
-              nullify(f)
-            end if
-
-            if(associated(f_cursor % array)) then
-              deallocate(f_cursor % array)
-            end if
-
-            if ( associated(f_cursor % attLists) ) then
-               do i = 1, size(f_cursor % attLists, dim=1)
-                  call mpas_deallocate_attlist(f_cursor % attLists(i) % attList, iErr)
-               end do
-               deallocate(f_cursor % attLists)
-            end if
+           f_cursor => f
+           do while(associated(f_cursor))
+               f_next => f_cursor % next
+
+               if (associated(f_cursor % array)) then
+                   deallocate(f_cursor % array)
+               end if
+
+               if (associated(f_cursor % attLists)) then
+                   do i = 1, size(f_cursor % attLists, dim=1)
+                       call mpas_deallocate_attlist(f_cursor % attLists(i) % attList, iErr)
+                   end do
+                   deallocate(f_cursor % attLists)
+               end if
+
+               if (associated(f_cursor % constituentNames)) then
+                   deallocate(f_cursor % constituentNames)
+               end if
 
-            deallocate(f_cursor)
+               if ((.not. local_keep_head_ptr) .or. (.not. associated(f_cursor, f))) then
+                   deallocate(f_cursor)
```

Comment:
@mgduda I did a local test merge into ocean/develop. I get a seg fault here when I compile with `intel/17.0.1` in debug mode, also with gnu, but only if I run multiple blocks per core. That is, if the flag `config_number_of_blocks` is set to a different number from the `mpirun -n` number of cores.  For example,
```
config_number_of_blocks=8
mpirun -n 4 ./ocean_model
```
causes a seg fault, but 
```
config_number_of_blocks=8
mpirun -n 8 ./ocean_model
```
runs fine. Using `config_number_of_blocks=0` is also fine.